### ISSUE TITLE
Out of Memory after a few days of usage

### ISSUE THREAD

**Issue by jsmwoolf** [FusionAuth/fusionauth-issues#111]
### Description
After running FusionAuth for a few days, the following log message is created:
```
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 12288 bytes for committing reserved memory.
# Possible reasons:
#   The system is out of physical RAM or swap space
#   In 32 bit mode, the process size limit was hit
# Possible solutions:
#   Reduce memory load on the system
```

I know that you can just reboot the machine and things will work again.  However, I want to know about a better way of preventing this issue in a production environment.

**Comment by robotdan** [comment]
It looks like the underlying host was unable to allocate the memory assigned to the VM.  This is generally not a Java / FusionAuth issue, but related to the resources on the host system. 

How much RAM have you assigned to the FusionAuth service? `fusionauth-app.memory` or `FUSIONAUTH_MEMORY` is using an environment variable in Docker or K8s. 
https://fusionauth.io/docs/v1/tech/reference/configuration
https://fusionauth.io/docs/v1/tech/installation-guide/docker

In a simplified view of the VM, if you assign `fusionauth-app.memory=1G` , the VM will not allocate all of this right away. Instead it will allocate memory as needed up to 1G. But if the system gets busy and you have other services using the RAM, then when the VM goes to allocate additional memory and none is available, it can fail as you have seen. 

You can ensure the host has enough RAM, and if you want to ensure the Java VM allocates all of the assigned RAM at startup you can pass `-XX:+AlwaysPreTouch` to the VM. This will slow down the VM startup as it will "touch" or allocate each memory page on start instead of dynamically increasing the heap size.  I don't know if this will always solve the issue you have seen since, it may depend upon if the VM is requesting additional heap allocation or something else. 

Hope that helps.

**Comment by botnyx** [comment]
As im planning to use FusionAuth for the long term, i thought it might be handy to drop my configuration here.

Initially, im running FusionAuth on Ubuntu 18, in a KVM as minimal as possible,
The VM has very little memory, which should be a good place to start imho.

![image](https://user-images.githubusercontent.com/27279013/56093001-e24f6b00-5ec3-11e9-8236-81b0f3ad5e0e.png)

## My FusionAuth Configuration
database: Postgres
fusionauth-app.memory=256M
fusionauth-search.memory=256M

At the moment, im not passing extra java-options.


**@jsmwoolf i am curious,** how many logins/registrations do you have while expiriencing this?  if possible, can you summarize what's on your  '/admin/report/totals' page?

i am still testing, so my stats so low, its no use to post them.  (~200 logins per day)

**Comment by jsmwoolf** [comment]
@botnyx I only have a few registrations since this is in a test environment.  The system that I'm running on is a t2.micro EC2 instance on AWS.

**Comment by robotdan** [comment]
Closing, not much we can do when the host VM runs out of memory. If you think you're running into a FusionAuth issue let us know! Thanks.
